Fixes race between starting transaction and shutdown

tinwelint opened the pull request:
This was designed to work already, but there was a slight gap in that
armor. DatabaseAvailability is the among last components to register in
the database lifecycle and therefore one of the first to be stopped.
Stopping DatabaseAvailability will raise the availability guard and then
await all active transactions to be finished (with some configurable
timeout). Starting transactions also checked the availability guard
state and registered themselves as active.

The problem was that the active state wasn't registered until after the
availability guard had been checked and so this little window between
those calls was racing with that same window in DatabaseAvailability
potentially allowing a transaction to start without being aware that
database was shutting down and shutdown process being unaware of an
active transaction running concurrently. This could result in
unspecified exceptions being thrown due to transaction running inside
components which were shutting down or even had been shut down.

This commit tightens this gap by letting transaction starting register
itself as active before checking availability guard. This way
transactions starting concurrently with shutdown will deterministically
succeed or fail, all depending on whether they made it in before the
availability guard was raised for shutdown.

This still leaves the max timeout that DatabaseAvailability waits before
going ahead with the shut down anyway. But that is by default a timeout
reasonable for most setups and it's configurable.

tinwelint committed:
A bit simpler to inject custom components into TestGraphDatabaseFactory

tinwelint committed:
Fixes race between starting transaction and shutdown

This was designed to work already, but there was a slight gap in that
armor. DatabaseAvailability is the among last components to register in
the database lifecycle and therefore one of the first to be stopped.
Stopping DatabaseAvailability will raise the availability guard and then
await all active transactions to be finished (with some configurable
timeout). Starting transactions also checked the availability guard
state and registered themselves as active.

The problem was that the active state wasn't registered until after the
availability guard had been checked and so this little window between
those calls was racing with that same window in DatabaseAvailability
potentially allowing a transaction to start without being aware that
database was shutting down and shutdown process being unaware of an
active transaction running concurrently. This could result in
unspecified exceptions being thrown due to transaction running inside
components which were shutting down or even had been shut down.

This commit tightens this gap by letting transaction starting register
itself as active before checking availability guard. This way
transactions starting concurrently with shutdown will deterministically
succeed or fail, all depending on whether they made it in before the
availability guard was raised for shutdown.

This still leaves the max timeout that DatabaseAvailability waits before
going ahead with the shut down anyway. But that is by default a timeout
reasonable for most setups and it's configurable.

tinwelint committed:
More convenient test utilities around PageCacheRule and NeoStoreDataSourceRule

chrisvest commented on community/io/src/test/java/org/neo4j/test/rule/PageCacheAndDependenciesRule.java:
```diff
@@ -0,0 +1,92 @@
+ * This is tedious to write and apply in the correct order in every test doing this. This rule collects
+ * this threesome into one rule for convenience.
+ */
+public class PageCacheAndDependenciesRule implements TestRule
```
👏 👏 👏

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/Kernel.java:
```diff
@@ -88,9 +96,21 @@ public KernelTransaction newTransaction( KernelTransaction.Type type, SecurityCo
+            transactionMonitor.transactionFinished( false, false );
+            throw new TransactionFailureException( Status.Transaction.TransactionStartFailed, e, e.getMessage() );
+        }
+        return transactions.newInstance( type, securityContext, timeout );
```
👍 Good comment, too.

chrisvest commented on community/kernel/src/test/java/org/neo4j/kernel/impl/api/KernelTest.java:
```diff
@@ -72,6 +97,28 @@ public void shouldNotAllowCreationOfConstraintsWhenInHA() throws Exception
+        {
+            // THEN
+            InOrder order = inOrder( transactionMonitor, availabilityGuard );
+            order.verify( transactionMonitor, times( 1 ) ).transactionStarted();
```
You forgot to verify the call to the availability guard.

tinwelint replied:
Oh indeed

tinwelint replied:
Now it's added

chrisvest reviewed (changes requested):
Looks good, I just want that test to be fixed.

burqen commented on community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/CoreAPIAvailabilityGuard.java:
```diff
@@ -50,11 +50,16 @@ public void assertDatabaseAvailable()
+    {
+        if ( guard.isShutdown() )
+        {
+            return new DatabaseShutdownException();
```
Should we not add initial exception as suppressed?

chrisvest commented:
@tinwelint There's a weird test failure.

tinwelint commented:
Looks related actually

tinwelint committed:
Moves AvailabilityGuard#shutdown() call back to first thing in db shutdown

Because it acts as a flag signalling to others that the db is shutting
down. This was otherwise a problem in some loops that would continue
until a database was shutting down.

tinwelint commented:
@chrisvest this has now been fixed... should go green

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactions.java:
```diff
@@ -301,10 +300,6 @@ KernelTransactionHandle createHandle( KernelTransactionImplementation tx )
-        if ( availabilityGuard.isShutdown() )
-        {
-            throw new DatabaseShutdownException();
-        }
```
I think `KernelTransactionsTest.exceptionWhenStartingNewTransactionOnShutdownInstance` is failing because of this change.

tinwelint committed:
Fixes unit test for shutdown check change

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/api/KernelTransactions.java:
```diff
@@ -301,10 +300,6 @@ KernelTransactionHandle createHandle( KernelTransactionImplementation tx )
 
     private void assertRunning()
     {
-        if ( availabilityGuard.isShutdown() )
```
this check was done on this layer before to ensure that we hold newTransactionsLock when we doing so, now in cases when we block for various reasons on this lock for example and wake up, later on, we will allow the transaction to be started even db is already off do not think that is desired behaviour

jastax commented:
Will this fix (some of) [9121](https://github.com/neo4j/neo4j/issues/9121)? 

I know you guys are incredibly busy and doing great work, but I need at least some kind of comment on the issue (as in: we have noticed, will look into it)

I'm running a system with a few thousand daily users and have to make a decision on how to go forward as the production database keeps getting killed every few weeks and won't recover without a full restart (and even that is not enough sometimes).

systay commented:
@tinwelint Any updates to this PR?

ragadeeshu commented:
@systay This will be updated shortly

ragadeeshu committed:
Let transactions fail gracefully on shutdown

We do this by checking if the database was shutdown while we were waiting for the transaction lock.

ragadeeshu commented:
@MishaDemianenko updated :)

MishaDemianenko reviewed (approved):
@ragadeeshu do not have questions anymore, thx :)
